### PR TITLE
[WIP] Gopherjs support

### PR DESCRIPTION
**Pull request by otraore**
This is still a work in progress and requires a lot of review and improvements.

### CODE REVIEW

File: assets_desktop.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+func (r *Region) Texture() *gl.Texture {
+	return r.texture.id
+}
+
+func (t *Texture) Texture() *gl.Texture {
+	return t.id
+}
+
+func LoadShader(vertSrc, fragSrc string) *gl.Program {
```
Comment (EtienneBruines) [inline assets_desktop.go]: Should be moved to below `NewTexture`.

File: assets_js.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		}
+
+	}
+
+	return &Texture{id, float32(img.Width()), float32(img.Height())}
+}
+
+func (t *Texture) Texture() *js.Object {
+	return t.id
+}
```
Comment (EtienneBruines) [inline assets_js.go]: The ordering of all definitions should match that of the _desktop file (looks better).

File: audio_js.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// +build netgo
+
+package engo
+
+import (
+	"log"
+
+	"engo.io/ecs"
+	"io"
```
Comment (EtienneBruines) [inline audio_js.go]: Should be near to "log", and not near "engo.io/ecs".

File: camera.go [hunk 4]
```diff
@@ -268,9 +269,11 @@ type EdgeScroller struct {
 func (*EdgeScroller) Priority() int          { return 10 }
 func (*EdgeScroller) Remove(ecs.BasicEntity) {}
 
+// TODO: Warning doesn't get the cursor position
 func (c *EdgeScroller) Update(dt float32) {
-	curX, curY := window.GetCursorPos()
-	maxX, maxY := window.GetSize()
+	curX, curY := 0.0, 0.0
+	maxX := dom.GetWindow().InnerHeight()
+	maxY := dom.GetWindow().InnerWidth()
```
Comment (EtienneBruines) [inline camera.go]: This replaces code; shouldn't we change the `window.GetSize()` methods instead?

File: demos/pong/pong.go [hunk 5]
```diff
@@ -84,14 +84,14 @@ func (pong *PongGame) Setup(w *ecs.World) {
 	}
 
 	score := Score{BasicEntity: ecs.NewBasic()}
-	score.RenderComponent = engo.NewRenderComponent(basicFont.Render(" "), engo.Point{1, 1}, "YOLO <3")
+	//score.RenderComponent = engo.NewRenderComponent(basicFont.Render(" "), engo.Point{1, 1}, "YOLO <3")
```
Comment (EtienneBruines) [inline demos/pong/pong.go]: This didn't work? Or errored?

File: engo_glfw.go [hunk 6]
```diff
@@ -23,15 +24,15 @@ import (
...
-	VResize            *glfw.Cursor
-	defaultCloseAction bool
-	close              bool
+	Arrow     *glfw.Cursor
+	IBeam     *glfw.Cursor
+	Crosshair *glfw.Cursor
+	Hand      *glfw.Cursor
+	HResize   *glfw.Cursor
+	VResize   *glfw.Cursor
+	close     bool
```
Comment (EtienneBruines) [inline engo_glfw.go]: Is `close` desktop-specific?

File: engo_js.go [hunk 7]
```diff
@@ -1,165 +1,126 @@
...
+	"image/color"
+	"io/ioutil"
 	"log"
 	"math"
 	"math/rand"
 	"strconv"
+	"time"
 
-	"engo.io/webgl"
+	"compress/gzip"
```
Comment (EtienneBruines) [inline engo_js.go]: compress/gzip should be with the others.

File: demos/pong/pong.go (hunk 5 above)
Comment (otraore) [inline demos/pong/pong.go]: Font is net fully supported yet (render is left)

File: engo_js.go [hunk 8]
```diff
@@ -1,165 +1,126 @@
...
+	WorldBounds.Max = Point{Width(), Height()}
+	Files.Load(func() {
+		log.Println("Assets loaded")
+	})
 
-	js.Global.Call("addEventListener", "keydown", func(ev js.Object) {
-		key := Key(ev.Get("keyCode").Int())
-		keyStates[key] = true
-	}, false)
+	log.Println("Hello Web")
```
Comment (EtienneBruines) [inline engo_js.go]: Debug log

File: engo_js.go [hunk 9]
```diff
@@ -1,165 +1,126 @@
...
+		keyStates[Key(ke.KeyCode)] = false
+	})
 
-	js.Global.Call("addEventListener", "keypress", func(ev js.Object) {
-		responder.Type(rune(ev.Get("charCode").Int()))
-	}, false)
+	Files = NewLoader()
+	WorldBounds.Max = Point{Width(), Height()}
+	Files.Load(func() {
+		log.Println("Assets loaded")
```
Comment (EtienneBruines) [inline engo_js.go]: Debug log. Also: not sure if we log this at _desktop? Let's be consistent.

File: engo_js.go [hunk 10]
```diff
@@ -191,29 +152,79 @@ func rafPolyfill() {
...
 		})
 	}
 }
 
-func RequestAnimationFrame(callback func(float32)) int {
+func RunIteration() {
+	keysUpdate()
+	currentWorld.Update(Time.Delta())
+	Time.Tick()
+	// TODO: this may not work, and sky-rocket the FPS
```
Comment (EtienneBruines) [inline engo_js.go]: I thought it would, but if it's anything like I wrote a while back; it runs at 60fps.

File: engo_js.go [hunk 11]
```diff
@@ -222,36 +233,64 @@ func loadImage(r Resource) (Image, error) {
...
+	// err := <-ch
+	// if err != nil {
+	// 	return "", err
+	// }
+
+	// return req.Get("responseText").Str(), nil
+}
 
-	return req.Get("responseText").Str(), nil
+func loadFont(r Resource) (*truetype.Font, error) {
```
Comment (EtienneBruines) [inline engo_js.go]: I hear you said that fonts were still a WIP? Is this part functional yet?

File: font_js.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
+// +build netgo
+
+package engo
+
+//TODO go generate to convert fonts to ones readable by engo
```
Comment (EtienneBruines) [inline font_js.go]: readable by engo? Truetype is relatively readable?

File: math.go [hunk 13]
```diff
@@ -8,6 +8,21 @@ type Point struct {
 	X, Y float32
 }
 
+// Abs returns the absolute value of x.
+//
+// Special cases are:
+//	Abs(±Inf) = +Inf
+//	Abs(NaN) = NaN
+func Abs(x float32) float32 {
```
Comment (EtienneBruines) [inline math.go]: We could also just use `luxengine/math`? Otherwise we'd have to implement these methods as we use them, anyway. And there's no downside to using `luxengine/math`?

File: render_js.go [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+	scale  Point
+	Color  color.Color
+	shader Shader
+	zIndex float32
+
+	drawable      Drawable
+	buffer        *js.Object
+	bufferContent []float32
+}
+
```
Comment (EtienneBruines) [inline render_js.go]: Looks like there's still a lot commented. Does this work even though it's commented?

File: math.go (hunk 13 above)
Comment (otraore) [inline math.go]: Gopherjs thought it was a native function and freaked out "math.Abs()"

File: shaders_js.go [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+package engo
+
+import (
+	"github.com/gopherjs/gopherjs/js"
+)
+
+type Shader interface {
+	Initialize(width, height float32)
+	Pre()
+	Draw(texture *js.Object, buffer *js.Object, x, y, rotation float32)
```
Comment (EtienneBruines) [inline shaders_js.go]: This would be where our `EngoEngine/gl` package would come in handy. We'd use `gl.Texture` here, and what it is beneath the hood isn't relevant to us. For the desktop, it's a `uint32`. For the web, it's `*js.Object`. 

I feel like it'd be nicer to use that, so we won't have to split the shaders in two. (because I don't see any other differences? )

File: assets.go [hunk 16]
```diff
@@ -165,31 +164,6 @@ func (l *Loader) Load(onFinish func()) {
 	onFinish()
 }
 
-type Image interface {
-	Data() interface{}
-	Width() int
-	Height() int
-}
-
-func LoadShader(vertSrc, fragSrc string) *gl.Program {
```
Comment (EtienneBruines) [inline assets.go]: See https://github.com/EngoEngine/engo/pull/247/files#r60846631 as well. 

This method should be able to run on the web? Since the `EngoEngine/gl` package has build stuff for netgo?

File: font_js.go (hunk 12 above)
Comment (otraore) [inline font_js.go]: https://github.com/zach-klippenstein/gofontviewer

File: shaders_js.go (hunk 15 above)
Comment (otraore) [inline shaders_js.go]: the webgl package won't accept gl.Texture it only accepts js.Object

Comment (EtienneBruines) [inline shaders_js.go]: It's `js.Object` under the hood, but the webgl *defines* `Texture`. 

https://github.com/EngoEngine/gl/blob/master/gl_webgl.go#L16

This is `EngoEngine.gl.Texture`. Which is what I am referring to.

Comment (otraore) [inline shaders_js.go]: I tried before and got this: pong.js:1 engo/render.go:259:25: cannot use e.RenderComponent.drawable.Texture() (value of type *github.com/gopherjs/gopherjs/js.Object) as *engo.io/gl.Texture value in argument to rs.currentShader.Draw

Maybe  I did it wrong?

Comment (EtienneBruines) [inline shaders_js.go]: You have to use "EngoEngine.gl.Texture" in the definition of the drawable / texture as well. In the comment I made on the top.

File: assets_js.go [hunk 17]
```diff
@@ -0,0 +1,68 @@
+// +build netgo
+
+package engo
+
+import (
+	"github.com/gopherjs/gopherjs/js"
+)
+
+type Image interface {
+	Data() *js.Object
```
Comment (EtienneBruines) [inline assets_js.go]: Here! :D

File: assets_js.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+	return program
+}
+
+func (r *Region) Texture() *js.Object {
+	return r.texture.id
+}
+
+type Texture struct {
+	id     *js.Object
```
Comment (EtienneBruines) [inline assets_js.go]: And here, and everywhere similar.

File: shaders_js.go (hunk 15 above)
Comment (otraore) [inline shaders_js.go]: Still can't get it to work, maybe you can try locally and tell me how it goes.

Comment (EtienneBruines) [inline shaders_js.go]: Have to leave in a few minutes, so I won't be able to help the next few hours. After that, sure thing.

File: font_js.go (hunk 12 above)
Comment (EtienneBruines) [inline font_js.go]: That's pretty much what we do already. So "copying" that to gopherjs wouldn't be that bad.

File: collision.go [hunk 19]
```diff
@@ -4,7 +4,6 @@ import (
 	"log"
 
 	"engo.io/ecs"
-	"github.com/luxengine/math"
```
Comment (paked) [inline collision.go]: Why did you remove luxengine/math, I thought we standardised on it?

Comment (EtienneBruines) [inline collision.go]: https://github.com/EngoEngine/engo/pull/247#discussion_r60846603

File: math.go (hunk 13 above)
Comment (paked) [inline math.go]: What if you aliased the import to `math32`?

File: engo_js.go [hunk 20]
```diff
@@ -191,29 +153,79 @@ func rafPolyfill() {
...
+	dom.GetWindow().AddEventListener("onbeforeunload", false, func(e dom.Event) {
+		dom.GetWindow().Alert("You're closing")
+	})
+}
+
+func runHeadless(defaultScene Scene) {
+	runLoop(defaultScene, true)
+}
+
+func runLoop(defaultScene Scene, headless bool) {
```
Comment (EtienneBruines) [inline engo_js.go]: This entire method isn't OS-specific.

File: engo_js.go [hunk 21]
```diff
@@ -191,29 +153,79 @@ func rafPolyfill() {
...
+func RunPreparation() {
+	keyStates = make(map[Key]bool)
+	Time = NewClock()
+
+	dom.GetWindow().AddEventListener("onbeforeunload", false, func(e dom.Event) {
+		dom.GetWindow().Alert("You're closing")
+	})
+}
+
+func runHeadless(defaultScene Scene) {
```
Comment (EtienneBruines) [inline engo_js.go]: This entire method isn't OS-specific.

File: engo_js.go [hunk 22]
```diff
@@ -1,165 +1,127 @@
...
-		keyStates[key] = false
-		// responder.Key(Key(ev.Get("keyCode").Int()), 0, RELEASE)
-	}, false)
+func SetBackground(c color.Color) {
+	if !headless {
+		r, g, b, a := c.RGBA()
+		Gl.ClearColor(float32(r), float32(g), float32(b), float32(a))
+		Gl.Clear(Gl.COLOR_BUFFER_BIT)
+	}
+}
```
Comment (EtienneBruines) [inline engo_js.go]: This isn't OS-specific either? If the `Gl.Clear` is required, we should also do it at the desktop-variant. 

I'd suggest moving this back to `engo.go`, because it's not OS-specific.

File: collision.go [hunk 23]
```diff
@@ -4,6 +4,7 @@ import (
 	"log"
 
 	"engo.io/ecs"
+	"github.com/luxengine/math"
```
Comment (EtienneBruines) [inline collision.go]: This error is related to the fact that `luxengine` uses fancy compiler-stuff for amd64 and i386 with the `Abs` method (the rest is fine). This is not related to the import path or function naming, but the implementation not being available for gopherjs.

Comment (EtienneBruines) [inline collision.go]: # Fixed

Simply `go get -u github.com/luxengine/math` to update your `math` library. It was already fixed before I could file a bug report :smile: .

**Comment by EtienneBruines** [comment]
## Summary current state
( @otraore can tell me if I missed anything ):

* Images can be loaded (using the direct filepath)
* Pong works, except for the score, because
* Fonts dont work at all, and result in crashes. This is mostly because it's not yet implemented. 
* Audio is not supported. 
* Some methods may be moved to `engo.go` because they're generic. (up for discussion - @paked ? )

There don't seem to be any strange changes in this PR anymore.

**Comment by otraore** [comment]
Really nice work @EtienneBruines, I still have to review all of your changes and I plan on adding in adding in Audio support soon as well.

**Comment by EtienneBruines** [comment]
# Summary

We are merging this branch, even though there are some small details left to fix. This is intentional. 

Please create an issue if you think there's anything seriously wrong with it. 

Thanks to @otraore and other contributors - I have extensively reviewed this PR, and it looks good to me. (Travis agrees :smile: )

**Comment by otraore** [comment]
> LGTM :) @EtienneBruines in the gitter chat

 🎉

**Comment by mewmew** [comment]
Oh, wow! Great work @otraore : )

**Comment by otraore** [comment]
@mewmew Thanks, thanks to @EtienneBruines @paked and every other contributor!!!
